### PR TITLE
Add support for hash and url configs to be used in connected_to

### PR DESCRIPTION
**Pull request by gmcgibbon**
### Summary

I noticed when using the new connection switching API, if I use a string database name, it bubbles down to [this connection resolution method](https://github.com/rails/rails/blob/8df7ed3b88fc9f19446d7207a745a331893b81cd/activerecord/lib/active_record/connection_adapters/connection_specification.rb#L216) which thinks strings are database URLs. Do we want to support URLs, or is this acceptable behaviour? If not, we should make `DatabaseConfig#spec_name` return a symbol instead of a string.

### CODE REVIEW

**Comment by gmcgibbon** [comment]
r? @eileencodes 
cc @rafaelfranca

**Comment by gmcgibbon** [comment]
Ok, I assumed it was different because you can't currently use a connection hash because of this [`to_sym` call](https://github.com/rails/rails/blob/8df7ed3b88fc9f19446d7207a745a331893b81cd/activerecord/lib/active_record/connection_handling.rb#L113). What do you think of `DatabaseConfig#spec_name` returning a symbol then?

**Comment by eileencodes** [comment]
Ohhhh I see - I was thinking about `connects_to`. Yes in that case we should remove the `to_sym` call in `connected_to`. That was a mistake on my part. Can you update this PR to remove that and add a test for a URL configuration? Thanks!

**Comment by eileencodes** [comment]
The underlying problem is that there are a billion ways to make a configuration and connect to the database. Take a look at this test https://github.com/rails/rails/blob/master/activerecord/test/cases/connection_adapters/merge_and_resolve_default_url_config_test.rb and you'll get an idea of some of the possible scenarios. We may eventually want to say some of these are not going to be possible with a multi-db application but I don't think we're there yet. For example, I'm not sure how a DATABASE_URL would work with multiple databases - since you can only pass one at a time.

**Comment by gmcgibbon** [comment]
Done! I added a test for both hash configs and URLs since they both fail on master.

File: activerecord/test/cases/connection_adapters/connection_handlers_multi_db_test.rb [hunk 1]
```diff
@@ -120,6 +120,33 @@ def test_switching_connections_via_handler
...
+          ENV["DATABASE_URL"] = previous_url
+        end
+
+        def test_switching_connections_with_database_config
+          previous_env, ENV["RAILS_ENV"] = ENV["RAILS_ENV"], "default_env"
+          config = { "adapter" => "sqlite3", "database" => "db/readonly.sqlite3" }
+
+          ActiveRecord::Base.connected_to(database: config) do
+            handler = ActiveRecord::Base.connection_handler
+            assert_equal handler, ActiveRecord::Base.connection_handlers[config]
```
Comment (gmcgibbon) [inline activerecord/test/cases/connection_adapters/connection_handlers_multi_db_test.rb]: I'm not sure how I feel about this. Should we try to lookup the config to store it as a hash key and fallback to this behaviour?

Comment (eileencodes) [reply activerecord/test/cases/connection_adapters/connection_handlers_multi_db_test.rb]: Oh 💩 it just occurred to me this isn't going to work. It's fine if it's a symbol but what would the handler be if it was a url? We definitely don't want the handler to be `postgres://blah` so I think we need to change connects to to take a role and a database at the same time and only if it's a symbol use the database name as a handler. Thoughts?

Comment (gmcgibbon) [reply activerecord/test/cases/connection_adapters/connection_handlers_multi_db_test.rb]: Yeah, I think if we always lookup handlers using symbols, that makes the most sense. Does it make sense to always use roles for handler lookup or do we need to still try using the database spec/env/config/url/??? for legacy support?

File: activerecord/lib/active_record/connection_handling.rb [hunk 2]
```diff
@@ -106,20 +106,20 @@ def connects_to(database: {})
...
     #   end
     def connected_to(database: nil, role: nil, &blk)
-      if database && role
-        raise ArgumentError, "connected_to can only accept a database or role argument, but not both arguments."
-      elsif database
+      if database.is_a?(Symbol) || database && role
+        hanlder_key = database.is_a?(Symbol) ? database : role
         config_hash = resolve_config_for_connection(database)
-        handler = lookup_connection_handler(database.to_sym)
+        handler = lookup_connection_handler(hanlder_key)
```
Comment (eileencodes) [inline activerecord/lib/active_record/connection_handling.rb]: You have a typo in handler 😄

Comment (gmcgibbon) [reply activerecord/lib/active_record/connection_handling.rb]: 🤦‍♂️

File: activerecord/test/cases/connection_adapters/connection_handlers_multi_db_test.rb (hunk 1 above)
Comment (gmcgibbon) [reply activerecord/test/cases/connection_adapters/connection_handlers_multi_db_test.rb]: > I've been thinking about this and I think we should either accept a symbol or a hash for the database option. If you need to pass a string then you need to also pass a role / name / handler with it.

I've refactored the method to behave this way for both URLs and config hashes. WDYT?

File: activerecord/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,18 @@
+*   Add support for hash and url configs in database hash of `ActiveRecord::Base.connected_to`.
```
Comment (bogdanvlviv) [inline activerecord/CHANGELOG.md]: Great addition. Thank you! Just want to note that we don't add changelog entries about complementing features that haven't been released.

**Pull request merged**